### DISCUSSION TITLE
Make FakeEncoder example Smart

### DISCUSSION THREAD

**Discussion by gringow831** [tttapa/Control-Surface#897] [Q&A]
Using the example of [FakeEncoder](#885 ), I have been trying to make is smart so it saves the potentiometer position from last use. My first thought was to save the value of potentiometer and retrieve it  when pressing and releasing the bank button or any other button. Also I thought I could maybe store the value last press and new one.
Anyone has any idea how to make this happen?

```ruby
Button number1 = 5;

int potValue1;
int potValue2;
int potValue3l;

void setup() {
    Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
  
  potValue1 = FakeEncoder; //potentiometer value
  
  byte valueHigh1 = highByte(potValue1);
  byte valueLow1 = lowByte(potValue1);

if (number1.update() == Button::Falling){

EEPROM.write(20, valueHigh1); //high value of FakeEncoder
EEPROM.write(21, valueLow1); //low value of FakeEncoder
    
  } else if (number1.getState() == Button::Rising){

    potValue1 = word(EEPROM.read(20),(21));
     }
```

I also thought I could save the value last time I press and retrieve it once released but not entirely sure if this is the way to go or if I am on the right path here. Then in some way to apply it to the banks too

```ruby
void loop() {
  Control_Surface.loop();
  potValue1 = analogRead(A0); //potentiometer value
    byte valueHigh1 = highByte(potValue1);
  byte valueLow1 = lowByte(potValue1);
if (number1.update() == Button::Falling){
EEPROM.write(20, valueHigh1); //high value of FakeEncoder
EEPROM.write(21, valueLow1); //low value of FakeEncoder
    
  } else if (number1.getState() == Button::Rising){
    auto activeBank = bank.getSelection();
  auto LastpotValue1 = word(EEPROM.read(20),(21));
  auto NewpotValue1 = analogRead(A0);
        potValue1 = word(EEPROM.read(20),(21));
    if (LastpotValue1 <= NewpotValue1){
      //wait for pot to go up and then move potValue
     };
    if (LastpotValue1 >= NewpotValue1){
      //wait for pot to go down and then move potValue
     };
     if (LastpotValue1 == NewpotValue1){
//if both are the same then move the potentiometer
     };
}}
```

EXAMPLE OF FAKE ENCODER
```ruby
#include <Control_Surface.h>

// Potentiometer that pretends to be a quadrature encoder
struct FakeEncoder {
    FakeEncoder(pin_t analogPin) : potentiometer(analogPin) {
    }
    FilteredAnalog<10> potentiometer;
    void begin() { potentiometer.setupADC(); }
    analog_t read() { return potentiometer.update(), potentiometer.getValue(); }
};

namespace CS { namespace Bankable { namespace ManyAddresses {

// Class that sends relative CC messages based on the position of a FakeEncoder
template <setting_t NumBanks>
struct FakeCCEncoder : GenericMIDIRotaryEncoder<FakeEncoder, ManyAddresses<NumBanks>, RelativeCCSender> {
    FakeCCEncoder(const Bank<NumBanks> &bank, FakeEncoder &&encoder, const Array<MIDIAddress, NumBanks> &addresses,
                  int16_t speedMultiply = 1, uint8_t pulsesPerStep = 1)
        : GenericMIDIRotaryEncoder<FakeEncoder, ManyAddresses<NumBanks>, RelativeCCSender>(
              {bank, addresses}, std::move(encoder), speedMultiply,
              pulsesPerStep, {}) {}
};

}}} // namespace CS::Bankable::ManyAddresses

Bankable::ManyAddresses::FakeCCEncoder<4> enc {bank, A0, {MCU::V_POT_1, MCU::V_POT_8, MCU::V_POT_3, MCU::V_POT_5}};

void setup() {
   Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```

**Comment by gringow831** [comment]
Anyone has any idea how to make this happen?

**Reply by Trickster-git** [reply]
It is interest question, I`m not sure about my idea.
If use Ableton and script it can possible, because on brand midi controllers when it turn on, hardwire pick up softwire values from Ableton.
My current though about that it no hardwire part, it DAW part. Because if you open first project, work on it, then you open secound project, hardwire need somehow remember positions. Now what happens if you use 200 encoders in controller and work on 20 projects?

**Reply by gringow831** [reply]
sure, I understand why saving on EEProm would be a bad choice. Any idea on how to adapt the script so it can be "smart" any other way?

**Reply by Trickster-git** [reply]
https://remotify.io/

And two source, I`m does not start with Ableton scripting, i try Fader Fox script in Ableton device mode.
[lambdacontrol-midi-remote-script-master.zip](https://github.com/tttapa/Control-Surface/files/10867272/lambdacontrol-midi-remote-script-master.zip)
[Faderfox_Universal_2(1).zip](https://github.com/tttapa/Control-Surface/files/10867273/Faderfox_Universal_2.1.zip)

You can search in YouTube old video how use Ableton scripting, it use Python.

**Reply by gringow831** [reply]
Thank you, didn't know the app but it won't work with the what I am making or the controller. 
Hence of trying to adapt that and make it smart if there is a way

**Comment by gringow831** [comment]
I am still trying to make this work. Do you have any idea I could use to remember the position of the pot in order to remember where to start from on the FakeEncoder @tttapa ?

**Reply by tttapa** [reply]
I'm sure it's possible, but it's unclear what the exact specifications/behavior are that you're after.

**Reply by gringow831** [reply]
Pretty much I am trying to make the FakeEncoder remember where it was left off as it's a relative value. so for example if I leave the FakeEncoder on 50%, change what it does with button and move lets say Breath to 75%. Once I go back to the FakeEncoder the 50% starts where the 75% of the Breath is rather than either starting at 75% or waiting till you hit the 50% of the movement with the FakeEncoder. Does it make sense?

Also, I have been checking the following example. Is there a way to use a potentiometer to do this? I tried with the CC14Potentiometer but it doesn't behave in the exact same way for some reason. 
```
void setup() {}

void loop() {
  const unsigned long interval = 100;
  static unsigned long prev_tick = millis() - interval;
  const uint16_t delta = 8;
  static uint16_t value = 0;
  const uint8_t address = 24;

  if (millis() - prev_tick > interval) {
    usbMIDI.sendControlChange(address + 0x00, (value >> 7) & 0x7f, 1);
    usbMIDI.sendControlChange(address + 0x20, (value >> 0) & 0x7f, 1);
    value += delta;
    if (value >= 16384)
      value = 0;
    prev_tick += interval;
  }
  while (usbMIDI.read());
}
```

**Comment by gringow831** [comment]
I think to explain the behaviour, I am trying to make a potentiometer when moving CW send 1 values and when moving CCW 65. 

I have tried a different way and I can send the 1 values but on the CCW I get 127. It would be interesting and better being able to use it though control_surface. do you think there is a way? I believe that way the controller will remember the value of potentiometer and there is no need to convert the potentiometer into FakeEncoder?
```
const int potPin = A0;
const int controllerNumber = 0x10; 
int lastPotValue = 0;

void setup() {
  pinMode(potPin, INPUT);
  lastPotValue = analogRead(potPin);
//  MIDI.begin(midiChannel);
}

void loop() {
  int potValue = analogRead(potPin);
  int delta = potValue - lastPotValue;
  if (abs(delta) > 2) {
    int midiValue = map(delta, -512, 512, -127, 127); 
    usbMIDI.sendControlChange(controllerNumber, midiValue, 1); 
    lastPotValue = potValue;
  }
  delay(10); 
}
```
